pkg/generate/app/componentresolvers.go

```diff
@@ -75,6 +75,20 @@ func (r PerfectMatchWeightedResolver) Resolve(value string) (*ComponentMatch, er
 	}
 	switch len(imperfect) {
 	case 0:
+		// If value is a file and there is a TemplateFileSearcher in one of the resolvers
+		// and trying to use it gives an error, use this error instead of ErrNoMatch.
+		// E.g., calling `oc new-app template.json` where template.json is a file
+		// with invalid JSON, it's better to return the JSON syntax error than a more
+		// generic message.
+		if isFile(value) {
```

Comment:
Come up with a new error type that resolves can guarantee users will get
(or a side error that will result in it being shown).

On Thu, Sep 17, 2015 at 10:58 AM, Rodolfo Carvalho <notifications@github.com
> wrote:

> In pkg/generate/app/componentresolvers.go
> <https://github.com/openshift/origin/pull/4660#discussion_r39755444>:
>
> > @@ -75,6 +75,20 @@ func (r PerfectMatchWeightedResolver) Resolve(value string) (*ComponentMatch, er
> >  	}
> >  	switch len(imperfect) {
> >  	case 0:
> > +		// If value is a file and there is a TemplateFileSearcher in one of the resolvers
> > +		// and trying to use it gives an error, use this error instead of ErrNoMatch.
> > +		// E.g., calling `oc new-app template.json` where template.json is a file
> > +		// with invalid JSON, it's better to return the JSON syntax error than a more
> > +		// generic message.
> > +		if isFile(value) {
>
> why doesn't template file searcher properly handle this and return an
> appropriate error?
>
> @smarterclayton <https://github.com/smarterclayton> TemplateFileSearcher
> does handle that and return an appropriate error. The problem is that we ignore
> all specific errors from resolvers
> <https://github.com/openshift/origin/blob/76a51df142c5299da4e4aee1272b87b67b1c429b/pkg/generate/app/componentresolvers.go#L71-L73>
> and return a generic ErrNoMatch that eventually became a poor error
> message as described in #4206
> <https://github.com/openshift/origin/issues/4206>.
>
> The way it works today is that if you have a file template.json with
> invalid data, running:
>
> $ oc new-app template.json
>
> will still succeed if e.g., there's a template named "template.json"
> stored in the server, or there's a Docker image named "template.json"...
> but when none of those things work, we were losing the relevant JSON
> SyntaxError...
>
> How do you think we could do this better?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/4660/files#r39755444>.
>



-- 
Clayton Coleman | Lead Engineer, OpenShift